comms/src/message/envelope.rs

```diff
@@ -100,6 +159,29 @@ impl MessageEnvelope {
         ))
     }
 
+    /// Modify and sign a forwarded MessageEnvelope
+    pub fn forward_construct(
+        node_identity: &NodeIdentity,
+        message_envelope: MessageEnvelope,
+    ) -> Result<Self, MessageError>
+    {
+        let mut message_envelope_header = message_envelope.deserialize_header()?;
+        message_envelope_header.peer_source = node_identity.identity.public_key.clone();
+        message_envelope_header.peer_signature = peer_signature(
+            node_identity,
+            message_envelope_header.origin_signature.clone(),
+            message_envelope.body_frame().clone(),
+        )?;
+
+        Ok(Self::new(
+            vec![WIRE_PROTOCOL_VERSION],
+            message_envelope_header
+                .to_binary()
+                .map_err(MessageError::MessageFormatError)?,
+            message_envelope.body_frame().clone(),
```

Comment:
You dont have to clone here, you take ownership of `message_envelope`.

Reply:
Unfortunately, it does need a copy due to a shared referencing. It might be possible at a later stage to reduce the number of copies by refactoring message envelope

Reply:
Agree - I've needed an `into_frames` elsewhere (for when you own message envelope and are done with it) - worth looking deeper into how we can reduce clones